File: 234. Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        '''
        O(n) time & space
        '''
        if not head or not head.next:
            return True
        ls = []
        while head:
            ls.append(head.val)
            head = head.next
        return ls == ls[::-1]

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        '''
        O(n) time and O(1) space
        '''
        if not head or not head.next:
            return True
        rev = None
        slow = fast = head
        while fast and fast.next:
            fast = fast.next.next
            temp = rev
            rev = slow
            slow = slow.next
            rev.next = temp
        if fast:
            slow = slow.next
        while rev and rev.val == slow.val:
            slow = slow.next
            rev = rev.next
        return not rev

File: 234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        ListNode node=head, prev=null, next=null, fast=head, slow=head;
        while (fast!=null && fast.next!=null) {
            fast = fast.next.next;
            slow = slow.next;
            next = node.next;
            node.next = prev;
            prev = node;
            node = next;
        }
        if (fast != null) {
            slow = slow.next;
        }
        while (slow != null) {
            if (slow.val != prev.val) {
                return false;
            }
            prev = prev.next;
            slow = slow.next;
        }
        return true;
    }
}